Add OSS/FOSS Lifecycle+Sustainability lesson

rockenman1234 opened the pull request:
**Hi everyone!**

Between projects, I took some time to create a new Python notebook that walks through FOSS lifecycle and sustainability practices. This aligns with the desired features outlined in [open issue #16](https://github.com/gt-ospo/oss-training/issues/16).

The notebook includes both Markdown explanations and custom Python code that I wrote to assess the sustainability of a repository by simply providing its URL. Here's a quick preview of the output:

![SCR-20250630-lwgs](https://github.com/user-attachments/assets/084cf76a-cba2-4724-b8d8-208e72effaf6)

Would love your feedback - thanks for taking a look! 😄

jyoung3131 commented:
Thanks, @rockenman1234! Let me review and get back to you this week.

jyoung3131 commented:
Some of these resources were shared as well which might be useful!

[6 Things You Can Do Right Now To Improve Your Open Source Project](https://zenodo.org/records/15678444)

[repo-scaffolder](https://github.com/DSACMS/repo-scaffolder/tree/main)

rockenman1234 commented:
@jyoung3131 I've added the new resources, please let me know what you think - and please feel free to make updates as needed!

Thank you! :-)

rockenman1234 commented:
I've added the new section on signoff and verified commits to `oss-module-01`, I didn't want to waste your time with another PR, so I added it here - please let me know if you want me to undo that!

rockenman1234 commented:
@jyoung3131 Tweaks have been made!

rockenman1234 commented on notebook-lessons/oss-module-08-foss-code-life.ipynb:
```diff
@@ -0,0 +1,370 @@
+    "    indicators = analyze_repo(repo_url)\n",
+    "    print(\"--- Sustainability Indicators ---\")\n",
+    "    for key, value in indicators.items():\n",
+    "        emoji = \"✅\" if value else \"❌\"\n",
```
Yeah I'm not too sure about this one, I don't think copilot knows much about how the GitHub API works?

rockenman1234 commented:
> Copilot encountered an error and was unable to review this pull request. You can try again by re-requesting a review.

I'll take that as "job well done" 😂

RonRahaman commented on notebook-lessons/oss-module-01-git.ipynb:
```diff
@@ -2709,6 +2712,246 @@
+    "\n",
+    "### What is `git commit --signoff`?\n",
+    "\n",
+    "The `--signoff` flag adds a \"Signed-off-by\" line to your commit message, which is a **requirement in some projects** - most notably the Linux kernel and official IBM projects.\n",
```
I suggest changing "most notably" to "for example"

RonRahaman commented on notebook-lessons/oss-module-01-git.ipynb:
```diff
@@ -2709,6 +2712,246 @@
+    "\n",
+    "The `--signoff` flag adds a \"Signed-off-by\" line to your commit message, which is a **requirement in some projects** - most notably the Linux kernel and official IBM projects.\n",
+    "\n",
+    "This practice originated in response to the [SCO lawsuit(s) versus IBM](https://en.wikipedia.org/wiki/SCO%E2%80%93Linux_disputes) and helps to solve concerns around copyright infringement. The sign-off effectively serves as a **Developer Certificate of Origin (DCO)**, asserting that:\n",
```
I suggest this instead.  There were several stakeholders involved in the series of lawsuits and countersuits

> In the Linux kernel, this practice originated from lawsuits and countersuits between [SCO lawsuit(s) and Linux vendors](https://en.wikipedia.org/wiki/SCO%E2%80%93Linux_disputes) and helps to resolve concerns around copyright infringement. The sign-off effectively serves as a **Developer Certificate of Origin (DCO)**, asserting that

RonRahaman commented on notebook-lessons/oss-module-01-git.ipynb:
````diff
@@ -2709,6 +2712,246 @@
+    "git commit -m \"Your commit message\"\n",
+    "```\n",
+    "\n",
+    "### Fixing Missing DCO Signoff\n",
+    "\n",
+    "If you forgot to sign off on past commits, you can retroactively apply the sign-off:\n",
+    "\n",
+    "```bash\n",
+    "# Note: replace the `#` with however many commits you want to go back and update\n",
+    "git rebase -i HEAD~# --signoff\n",
+    "\n",
+    "# Then run:\n",
+    "git push --force-with-lease origin main\n",
+    "```\n",
+    "\n",
+    "> **Warning**: This will replace your commit messages if you're not careful! Always backup your work before force-pushing."
````
I recommend omitting this section.  In collaborative workflows, there are many cases where (in general) force-pushing to a published branch can cause irreparable harm to the team.  You might think that it won't matter if you do it in an un-merged branch on your own fork, but it's not uncommon for other team members to base their changes off of your un-merged branch.  

So the situation for fixing missing signoffs is complicated, depends on your team's practices, and it's and best resolved by asking your project manager.  We shouldn't be overly-proscriptive in our guide.

RonRahaman commented on notebook-lessons/oss-module-01-git.ipynb:
````diff
@@ -2709,6 +2712,246 @@
+    "\n",
+    "3. **Repository-Specific Settings**: If you don't want to make these changes globally, omit the `--global` flag and run these commands in each repo where they should apply.\n",
+    "\n",
+    "4. **Fixing Commits**: If you need to fix a commit:\n",
+    "   ```bash\n",
+    "   git commit --amend -m \"Updated message\"  # Uses your existing config\n",
````
Either omit this or mention that `git commit --amend` is recommended only for unpublished changes (i.e. before you've pushed the changes)

RonRahaman commented on notebook-lessons/oss-module-08-foss-code-life.ipynb:
```diff
@@ -0,0 +1,391 @@
+    "\n",
+    "#### 🌐 Ecosystem Phase\n",
+    "- Community forms, contributions increase.\n",
+    "- Features grow, integrations appear.\n",
```
In my experience, integrations often appear in the project phase.  Many projects pursue integrations (CI/CD, documentation generations) in the early stages of the project.  They're helpful even when the team is small.

RonRahaman commented on notebook-lessons/oss-module-08-foss-code-life.ipynb:
```diff
@@ -0,0 +1,391 @@
+    "### 🌐 Ecosystem Phase\n",
+    "- **Actionable Steps**:\n",
+    "  - Foster community engagement through GitHub Discussions.\n",
+    "  - Automate workflows with GitHub Actions.\n",
```
See my comment above.  In my experience, GitHub Actions are often used in early in the project.  They are low-overhead and helpful even for small teams.

RonRahaman commented on notebook-lessons/oss-module-08-foss-code-life.ipynb:
```diff
@@ -0,0 +1,391 @@
+    "- ✅ CONTRIBUTING.md to guide newcomers\n",
+    "- ✅ GitHub Actions for CI/CD\n",
+    "- ✅ Good Issue and PR templates\n",
+    "- ✅ Documentation hosted with tools like MkDocs or ReadTheDocs\n",
```
MkDocs doesn't provide hosting capabilities, as far as I know.  Distinguish between documentation generation (mkdocs, sphinx, etc) and documentation hosting (readthedocs, GitHub pages)

RonRahaman commented on notebook-lessons/oss-module-08-foss-code-life.ipynb:
```diff
@@ -0,0 +1,391 @@
+    "---\n",
+    "\n",
+    "## 📋 6 Things You Can Do Right Now to Improve Your Open Source Project\n",
+    "*[Prepared by the OpenSource@Stanford team (Zach Chandler, Francesca Vera)](https://zenodo.org/records/15678444)*\n",
```
I'd make this a bit more prominent via the markdown formatting.

RonRahaman reviewed (changes requested):
@rockenman1234 this is amazing work overall, and thank you so much for putting the time into this.  Please see my suggested changes.

jyoung3131 commented:
Hi all - I've reviewed the changes and they seem ok. I'm going to merge this for today's session.
